libvirt/tests/src/virsh_cmd/domain/virsh_domdirtyrate_calc.py

```
@@ -41,18 +66,26 @@ def check_output():
         if out_dict["dirtyrate.calc_period"] != period:
             test.fail("Calculating period is not the same with "
                       "the setting period %s" % period)
+        if mode and out_dict["dirtyrate.calc_mode"] != mode:
+            test.fail("Calculating mode is not the same with "
+                      "the specified mode %s" % mode)
         dirty_rate = out_dict["dirtyrate.megabytes_per_second"]
-        if abs(int(dirty_rate)/int(ram_size) - 1) > 0.75:
-            test.fail("Dirty rate calculated %s has a big difference "
-                      "with the ram size %s loaded in guest " % (dirty_rate, ram_size))
+        if mode and mode in "dirty-ring":
+            for cpu_num in range(vm.get_cpu_count()):
+                dirty_rate = out_dict["dirtyrate.vcpu.%s.megabytes_per_second" % cpu_num]
+                check_dirty_rate(dirty_rate)
+        else:
+            check_dirty_rate(dirty_rate)
 
     vm_name = params.get("main_vm")
     status_error = "yes" == params.get("status_error", "no")
     option = params.get("option", " ")
+    mode = params.get("mode")
     period = params.get("period", "1")
     ram_size = params.get("ram_size")
     num_of_sec = params.get("num_of_sec", 10000)
     calc_status = params.get("calc_status", "2")
+    dirty_ring_size = params.get("dirty_ring_size")
 
     if not libvirt_version.is_libvirt_feature_supported(params):
         test.cancel("domdirtyrate-calc command is not supported "
```

Please notice this is not working. If you want to throw customized error message, you can define 'unsupported_err_msg' parameter for is_libvirt_feature_supported().

---

Updated. Thanks